packages/rest-crud/src/crud-rest-builder.plugin.ts

```diff
@@ -0,0 +1,123 @@
+// Copyright IBM Corp. 2020. All Rights Reserved.
+// Node module: @loopback/rest-crud
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {bind, ControllerClass, inject} from '@loopback/core';
+import {
+  asModelApiBuilder,
+  ModelApiBuilder,
+  ModelApiConfig,
+} from '@loopback/model-api-builder';
+import {
+  ApplicationWithRepositories,
+  Class,
+  Entity,
+  EntityCrudRepository,
+} from '@loopback/repository';
+import {Model} from '@loopback/rest';
+import debugFactory from 'debug';
+import {defineCrudRepositoryClass, defineCrudRestController} from '.';
+
+const debug = debugFactory('loopback:boot:crud-rest');
+
+export interface ModelCrudRestApiConfig extends ModelApiConfig {
+  // E.g. '/products'
+  basePath: string;
+}
+
+@bind(asModelApiBuilder)
+export class CrudRestApiBuilder implements ModelApiBuilder {
+  readonly pattern: string = 'CrudRest';
+
+  build(
+    application: ApplicationWithRepositories,
+    modelClass: typeof Model & {prototype: Model},
+    cfg: ModelApiConfig,
+  ): Promise<void> {
+    const modelName = modelClass.name;
+    const config = cfg as ModelCrudRestApiConfig;
+    if (!config.basePath) {
+      throw new Error(
+        `Missing required field "basePath" in configuration for model ${modelName}.`,
+      );
+    }
+
+    if (!(modelClass.prototype instanceof Entity)) {
+      throw new Error(
+        `CrudRestController requires an Entity, Models are not supported. (Model name: ${modelName})`,
+      );
+    }
+    const entityClass = modelClass as typeof Entity & {prototype: Entity};
+
+    let repoClassName = entityClass.name + 'Repository';
+
+    try {
+      application.getBinding('repositories.' + repoClassName);
+      debug(
+        'Using repository class',
+        repoClassName,
+        ', as it is already bound to application',
+      );
+    } catch {
+      // repository class does not exist
+      const repositoryClass = setupCrudRepository(entityClass, config);
+      application.repository(repositoryClass);
+      repoClassName = repositoryClass.name;
+      debug('Registered repository class', repoClassName);
+    }
+
+    const controllerClass = setupCrudRestController(entityClass, config);
+    application.controller(controllerClass);
+    debug('Registered controller class', controllerClass.name);
+
+    return Promise.resolve();
+  }
+}
+
+/**
+ * Set up a CRUD Repository class for the given Entity class.
+ *
+ * @param entityClass - the Entity class the repository is built for
+ * @param modelConfig - configuration of the Entity class
+ */
+function setupCrudRepository(
+  entityClass: typeof Entity & {prototype: Entity},
+  modelConfig: ModelCrudRestApiConfig,
+): Class<EntityCrudRepository<Entity, unknown>> {
+  const repositoryClass = defineCrudRepositoryClass(entityClass);
+
+  inject(`datasources.${modelConfig.dataSource}`)(
+    repositoryClass,
+    undefined,
+    0,
```

Comment:
@nabdelgadir might be a silly question: I am looking at the [`inject`](https://github.com/strongloop/loopback-next/blob/master/packages/context/src/inject.ts#L127) function, how do we know the index is 0 and why the `member` name is undefined?

Reply:
From my understanding, the only reason we're using a number (that happens to be 0 in this case) is so that it goes into [this `if` statement](https://github.com/strongloop/loopback-next/blob/master/packages/context/src/inject.ts#L134), but I don't think the number actually matters. And we're only passing in `member` since it's required and so we're able to pass in the number as the third argument, so leaving it as undefined is the easiest way to go about it (but I don't actually know the role of `member` in the injection, maybe @bajtos could comment on this?).

Reply:
@bajtos  ^ (I have the same question)

Reply:
When you set the member to `undefined`, it means the constructor function. (The constructor does not have any name.

The index `0` is identifying the first argument of the target function (the class constructor in this case).

In this particular invocation of `inject`, we are decorating the first argument of the repository constructor to inject the datasource instance. It's an equivalent of the following TypeScript code:

```ts
class MyRepository /*extends ...*/ {
  constructor(
    @inject(`datasources.${modelConfig.dataSource}`)
    public dataSource: juggler.DataSource
  ) {}
}
```

Because decorators are converted to JavaScript calls of TypeScript API at compile time, we cannot use decorator syntax when building the repository class dynamically at runtime, we have to invoke the decorator function manually.

Reply:
Can we create a constant and use that instead of 0? Perhaps also one for the undefined value? So both constants can provide some clarity to the method call?

```
inject(`datasources.${modelConfig.dataSource}`)(
    repositoryClass,
    UNDEFINED_CLASS_CONSTRUCTOR_FUNCTION,
    INDEX_OF_INJECTED_DATASOURCE_PARAMETER_INSIDE_REPOSITORY_CONSTRUCTOR,
```

I know the names are silly, but you get the point.

Reply:
> Because decorators are converted to JavaScript calls of TypeScript API at compile time, we cannot use decorator syntax when building the repository class dynamically at runtime, we have to invoke the decorator function manually.

Thanks everyone for the reply. This is something I didn't know before 👍